cpp/src/strings/search/find_instance.cu

```diff
@@ -0,0 +1,135 @@
+                                                string_view const d_target,
+                                                size_type const instance,
+                                                size_type* d_results)
+{
+  auto const tid     = cudf::detail::grid_1d::global_thread_id();
+  auto const str_idx = tid / cudf::detail::warp_size;
+  if (str_idx >= d_strings.size() or d_strings.is_null(str_idx)) { return; }
+
+  namespace cg        = cooperative_groups;
+  auto const warp     = cg::tiled_partition<cudf::detail::warp_size>(cg::this_thread_block());
+  auto const lane_idx = warp.thread_rank();
+
+  auto const d_str = d_strings.element<string_view>(str_idx);
+  auto const begin = d_str.data();
+  auto const end   = begin + d_str.size_bytes();
+
+  // each thread compares the target with the thread's individual starting byte for its string
+  auto const max_pos   = d_str.size_bytes();
+  size_type char_pos   = max_pos;
+  size_type char_count = 0;
+  size_type count      = 0;
+  for (auto itr = begin + lane_idx; itr + d_target.size_bytes() <= end;
+       itr += cudf::detail::warp_size) {
+    size_type const is_char = !is_utf8_continuation_char(*itr);
+    size_type const found   = is_char && (d_target.compare(itr, d_target.size_bytes()) == 0);
+    auto const found_count  = cg::reduce(warp, found, cg::plus<int>());
+    auto const found_scan   = cg::inclusive_scan(warp, found);
+    auto const chars_scan   = cg::exclusive_scan(warp, is_char);
+    auto const found_pos    = (found_scan + count) == (instance + 1) ? chars_scan : char_pos;
+    char_pos                = cg::reduce(warp, found_pos, cg::less<size_type>());
+    if (char_pos < max_pos) { break; }
+    count += found_count;
+    char_count += cg::reduce(warp, is_char, cg::plus<size_type>());
+  }
+
+  if (lane_idx == 0) { d_results[str_idx] = char_pos == max_pos ? -1 : char_pos + char_count; }
+}
+
+}  // namespace
+
+std::unique_ptr<column> find_instance(strings_column_view const& input,
+                                      string_scalar const& target,
+                                      size_type instance,
+                                      rmm::cuda_stream_view stream,
+                                      rmm::device_async_resource_ref mr)
+{
+  CUDF_EXPECTS(
+    instance >= 0, "Parameter instance must be positive integer or zero.", std::invalid_argument);
+  CUDF_EXPECTS(target.is_valid(stream), "Parameter target must be valid.", std::invalid_argument);
+
+  // create output column
+  auto results = make_numeric_column(data_type{type_to_id<size_type>()},
+                                     input.size(),
+                                     cudf::detail::copy_bitmask(input.parent(), stream, mr),
+                                     input.null_count(),
+                                     stream,
+                                     mr);
+  // if input is empty or all-null then we are done
+  if (input.size() == input.null_count()) { return results; }
```

Comment:
Aha, clever thinking here. This would also be true if `input.empty()` so no need to check again